### ISSUE TITLE
X Offset function button

### ISSUE THREAD

**Issue by joshnortheimer** [bitfocus/companion#2532]
### Is this a feature relevant to companion itself, and not a module?

- [X] I believe this to be a feature for companion, not a module

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the feature

in the surfaces tab, when you get the settings of a streamdeck, you can select the x and y offset, for all streamdecks that aren't XL. i would like a function added that i can assign to a button that allows a specific streamdeck's offset be changed.

### Usecases

I would use this, for instance, with a streamdeck +, which allows me to see 16 of the 32 buttons. but, i could make one button toggle the x offset from 1 to 5, and then another to set it back, so it would work like a page right/left button, so i could use all buttons on a page with a streamdeck+, thereby using fewer of the 99 pages of buttons. like this:
![x-offset button](https://github.com/bitfocus/companion/assets/42381830/a957e399-234b-4878-b28e-7dc2cbaace9f)

I could also then use a streamdeck + on pages designed for the XL, by adding the x-offset function to a dial on every page, so it wouldn't even take up button function on the XLs. i could also put multiple pages for a mini on one page.

**Comment by Julusian** [comment]
In a future version (3.2? 3.3?) it will be possible to both resize the grid and to have more than 99 pages. Will this still be useful once that is done?

> I could also then use a streamdeck + on pages designed for the XL, by adding the x-offset function to a dial on every page

I'm not sure about this use case, you likely wouldnt be able to utilise the dials fully as the page will need to make sense on an xl too, and the xl will loose an extra 2 controls for the offset adjustment.

**Comment by dnmeid** [comment]
And in a future, future version we'll have the windows and windows functionality will give you the same effect.
Not exactly offsetting the surface, but offsetting the cutout of controls you can see on a surface.
So I think this is a won't fix for this exact request, but we have this functionality on our roadmap.

**Comment by obcocav** [comment]
+1.  I had made a similar previous request in order to have one page show buttons designed for a 15 key deck on the XL in the lower right quadrant for ergonomics.

**Comment by joshnortheimer** [comment]
> In a future version (3.2? 3.3?) it will be possible to both resize the grid and to have more than 99 pages. Will this still be useful once that is done?

It would still be nice from a formatting perspective to have those controls grouped together, but having more than 99 pages definitely makes it less important.


> I'm not sure about this use case, you likely wouldnt be able to utilise the dials fully as the page will need to make sense on an xl too, and the xl will loose an extra 2 controls for the offset adjustment.

So, my thought would be that when viewing the page on an XL, you see the graphics and functions for pushing the button. on the +, turning a dial left or right would adjust the offset by 4, the button push function wouldn't change, so you wouldn't lose any real estate on the XL, you would pick up the extra function where needed on the +.

**Comment by dnmeid** [comment]
Hey guys, maybe you should have alook at #2387
With the windows system of this proposal all of your usecases will be possible and even more.
I guess we won't do any workarounds now just to change the behaviour later. So, please be patient.

**Comment by davidjoshuaford** [comment]
This would be useful!

+1 for Actions to alter a Surface's:

- Row count
- Column count
- Horizontal Offset in grid
- Vertical Offset in grid

Use case: I'm using a mini monitor as a teleprompter clock/ HUD, and would like to be able to toggle the emulator surface between a 2x7 layout and a 1x3 layout as well as shift the Offset to make it either a dashboard view or else a large clock. 

![Image](https://github.com/user-attachments/assets/2635d742-fe46-47d1-882e-3d6c6e3a8c19)


![Image](https://github.com/user-attachments/assets/0484c699-61e1-4063-ad2e-c18953c5363d)

**Comment by davidjoshuaford** [comment]
To add to the above, being able to 

- Set the horizontal/ vertical offset and count to an absolute row/column / count 
- Relatively increase/decrease by “x” the horizontal/vertical offset and count 
- (all of these accepting custom variables)

…would provide an alternate way to group or ungroup surfaces on the fly. (Instead of surfaces being permanently locked to a group in the surface tab) 

So for example, in a 4x16 button configuration layout you could have two XLs that can come together on one page for large button arrays (like input routing, audio routing, etc) with one surface set to 0/0 and the other to 0/8. Then when you jump away from that page they can return to solo surfaces realigned as 0/0 and 0/0. 

I wonder if this should be added into the “set surface to page” action that currently exists, or if it should be its own action.

**Comment by dnmeid** [comment]
@davidjoshuaford Let me qote myself.

> Hey guys, maybe you should have alook at [#2387](https://github.com/bitfocus/companion/discussions/2387) With the windows system of this proposal all of your usecases will be possible and even more. I guess we won't do any workarounds now just to change the behaviour later. So, please be patient.

**Comment by isaacdantzler** [comment]
Would love to have a UI sort of like what's going on with Buttons - I like the idea of moving a surface around a page rather than moving it to another page when I want "folder-like" actions.

**Comment by arikorn** [comment]
@Julusian, your solution is awesome, thank you! But... the inputs really need to be expressions, or at the very least, variables: Then the user can attach it to a rotation action -- on a second surface, like my Contour Shuttle ;) -- to get an "infinitely scrolling" set of buttons. (A colleague has also expressed this opinion, in the past, FWIW...).  If you're not planning on doing that, I could probably submit a PR for it. Just let me know.

**Comment by dnmeid** [comment]
@Julusian you see what happens ;-)
@arikorn unfortunately, our assessments diverge here. Offsetting a whole surface dynamically in my opinion is not a good idea.
What you want to do can be done with the surface API protocol. It is a workaround too, but a much more elegant one in terms of flexibility. Although the module is meant for bringing in buttons from remote machines, it also can be used for reusing buttons from the same installation and there you can assign the position of any button individually and it can use variables for the button target location.

There are a lot of people that prefer a more location based and less page based approach of organizing the currently visible buttons. I'm one of them and I think the backers off this request are too. So we should get this right.

**Comment by arikorn** [comment]
@Julusian, my sincerest apologies!!!  I seem to have overlooked something that should have been obvious. The "Surface: adjust position" action does exactly what I want, without the need for variables or expressions.

@dnmeid, thank you for your insights, as always. I would appreciate if you could take the time to reply to my PR #3505, or to my conversation with you on Slack on the same topic!  (I will add links in the PR to the recent updates I mentioned to you on Slack.)
